design/rich-authorization-requests.md

```diff
@@ -0,0 +1,173 @@
+
+Clients announce the authorization data types they use in the new dynamic client registration parameter "authorization_details_types".
+
+
+### Authorization Request
+
+As per spec,
+The `authorization_details` request parameter CAN be used to specify
+authorization requirements in all places where the "scope" parameter
+is used for the same purpose, examples include:
+
+*  Authorization requests as specified in [RFC6749],
+
+*  Access token requests as specified in [RFC6749], if also used as
+   authorization requests, e.g. in the case of assertion grant types
+   [RFC7521],
+
+*  Request objects as specified in [I-D.ietf-oauth-jwsreq],
+
+*  Device Authorization Request as specified in [RFC8628],
+
+*  Backchannel Authentication Requests as defined in [OpenID.CIBA].
+
+we will use it first in Authorization requests.
+
+### Token Response
+
+In addition to the token response parameters as defined in [RFC6749],
+the authorization server MUST also return the authorization details
+as granted by the resource owner and assigned to the respective
+access token.
+
+## Implementation details
+
+SPI
+
+````
+public interface RarProcessorProvider extends Provider {}
+
+public interface RarProcessorProviderFactory extends ProviderFactory<RarProcessorProvider> { }
+````
+e.g. implementation in extension
+
+````
+public class Psd2RarProcessorProvider implements RarProcessorProvider {}
+
+public class Psd2RarProcessorProviderFactory implements RarProcessorProviderFactory { 
+
+    public static final String ID = "rar-psd2";
+    public static final String SUPPORTED_DATA_TYPE = []; //Json array schema definition
+}
+````
+
+* Support advertisement of supported authorization details types in OAuth server metadata
+
+````
+OIDCConfigurationRepresentation config = new OIDCConfigurationRepresentation();
+config.setAuthorizationDataTypesSupported(
+    getSupportedAuthorizationDataTypes(RarProcessorProvider.class)
+    
+);
+````
+
+*  Check if authorization data type or authorization details is not conforming to the respective type
+   definition when authorization_details is available in the authorization requests 
+
+````
+rarProcessorProvider.isAuthorizationDetailsCompliant(authorizationDetails)
+````
+After the user authenticate, rarProcessor should:
+
+*  Determine how authorization details are shown to the user in the user consent
+*  Exchange with the resource server if necessary
+*  Enrich authorization details in the user consent process (e.g. add selected accounts or set expirations)
+*  show the consent page
+
+````
+rarProcessorProvider.prepareAuthorizationDetails()
+````
+   
+When the user submits the consent  
+
+*  Merge requested and pre-existing authorization_details (e.g. in case of update)  
+
+* store authorization_details within the grant
+
+````
+rarProcessorProvider.processAuthorizationDetails()
+````
+
+### Admin UI
+
+AS should allow the possibility to select RarProcessor Implementation Module.
+
+Files/Classes/methods affected:
+
```

Comment:
I think this should also list the adapter libraries that need to be updated in order to support the new `authorization_details` Authorization request parameter and the `authorization_details` Access token response attribute.

- Keycloak adapter-core (org.keycloak.adapters.OAuthRequestAuthenticator#authenticate)
- Keycloak JS Adapter

Reply:
Nice, I completely forgot about the adapters, I will add. However, I do not understand what needs to be changed in [org.keycloak.adapters.OAuthRequestAuthenticator#authenticate](https://github.com/keycloak/keycloak/blob/b1d39aa136662d54deca10c97374f0932ce6316b/adapters/oidc/adapter-core/src/main/java/org/keycloak/adapters/OAuthRequestAuthenticator.java#L263)